Feat/settings search

igorms-cons opened the pull request:
Feature: creating the search feature for settings.

Explanation:   Be able to search for a menu or section in all settings and navigate to it

Manual testing steps:  
  - Open Metamask 
  - Go to Settings 
  - type the section / tab (or its description) you want to go
  - Click et get navigate to it
  
Fix: https://github.com/MetaMask/metamask-extension/issues/12760

igorms-cons committed:
clean list search & fuse threshold decreased

igorms-cons committed:
update search-icon , fix tests

igorms-cons committed:
nice to have highlighting text & cleaning

igorms-cons committed:
unit test on settings & search input ui up on expanded view

igorms-cons committed:
fix error with color variable - fix rebase

igorms-cons committed:
fix color variable in alert scss

igorms-cons committed:
setting search input padding right up

igorms-cons commented:
Hello, here is a small video of feature. I recommended you to pull the branch and test it. 
The icons on search list are the same (the gear) because this [PR](https://github.com/MetaMask/metamask-extension/pull/12920) is not yet merged. (icons will be available then)


https://user-images.githubusercontent.com/89637087/149901736-68992c37-1dcd-460a-a2f4-4008971ddf91.mp4

igorms-cons committed:
fix dom warning

igorms-cons committed:
util/search test added & Dom element warning fix

igorms-cons committed:
renaming files

igorms-cons committed:
fix color text in settings search

igorms-cons committed:
settings search highlight text refacto & fix ui

igorms-cons committed:
fix settings-search test & renaming

mirjanaKukic commented:
Noticed that Search field does not look well on resolution from 575px-676px

![image](https://user-images.githubusercontent.com/94608179/151338542-474bfb17-59ba-48ee-a88b-04ea86836e7e.png)

![image](https://user-images.githubusercontent.com/94608179/151338785-30c2847d-db56-43c3-9603-d28cff6c7b23.png)

danjm commented on ui/pages/settings/settings-search-list/settings-search-list.js:
```diff
@@ -0,0 +1,113 @@
+import React, { Component } from 'react';
```
It would be best if the newly added components were functional components instead of class components. We can either fix that in this PR, or create a ticket to refactr that code, that is picked up immediately after this PR is merged

The relevant components are this one and `settings-search.js`

amerkadicE replied:
Done

mirjanaKukic commented:
I've investigated e2e tests which are failing and seems like there is an issue. While running tests it populates input search field with test data as it can be seen on screenshot : 

Test : creates first custom RPC entry

![image](https://user-images.githubusercontent.com/94608179/151808634-45a1f447-8786-4bee-88bf-e8209e199b3b.png)

It gets stuck since in test we are finding input type=text elements and now also search field is input text field.

Also tests: _warns user when they enter url for an already configured network_ and _warns user when they enter chainId for an already configured network_ are affected and failing.

danjm commented:
Nice find @mirjanaKukic

Relevant code is in lines 116-130 of `test/e2e/tests/custom-rpc-history.spec.js`

Commit:
Fix styling on search field for edge cases, update components and e2e

E2E tests update for search feature

Update components from class to functional component

		#

Fix storybook for search box

Fix styling

Fix unit tests

fix: remove z-index

Fix unit tests

danjm commented:
I think the logic here looks good. @georgewrmarshall Could you review this from a design system and css perspective? Are the patterns here roughly consistent with the direction we want to be heading?

igorms-cons commented:
> I think the logic here looks good. @georgewrmarshall Could you review this from a design system and css perspective? Are the patterns here roughly consistent with the direction we want to be heading?

@danjm  @amerkadicE  @david0xd  Don't forget about icons  I mentioned earlier

mirjanaKukic commented:
Verified by QA.

georgewrmarshall commented on ui/pages/settings/settings-search/settings-search.stories.js:
```diff
@@ -0,0 +1,11 @@
+import SettingsSearch from './settings-search';
+
+export default {
+  title: 'SettingsSearch',
```
The story title should echo the components location in the codebase

```suggestion
  title: 'Pages/Settings/SettingsSearch',
```

georgewrmarshall commented on ui/pages/settings/settings-search/settings-search.stories.js:
```diff
@@ -0,0 +1,11 @@
+
+export const SettingsSearchComponent = () => {
+  return <SettingsSearch />;
+};
```
Story also seems to be broken. Run `yarn storybook` and go to http://localhost:6007/?path=/story/ui-pages-settings-settings-search-settings-search-stories-js--settings-search-component to see the component 

<img width="1440" alt="Screen Shot 2022-02-05 at 2 48 25 PM" src="https://user-images.githubusercontent.com/8112138/152628366-0a7f6dad-6c9e-4151-a49d-460dcbdaddbc.png">

amerkadicE replied:
Done. 
You can see Search Box component in storybook now. 
Also if you want to use interactive Search Box you can do it through SettingsPage component in storybook.

georgewrmarshall replied:
Awesome thanks @amerkadicE!!

georgewrmarshall reviewed (changes requested):
Will do another review for the styles but could you try to get the story working in storybook please. You may need to pass the component some dummy data

amerkadicE commented on ui/components/ui/icon/search-icon.component.js:
```diff
@@ -0,0 +1,20 @@
+   * Color of the icon should be a valid design system color and is required
+   */
+  color: PropTypes.string.isRequired,
+};
```
Hi, @georgewrmarshall @yanrong-chen Did you decide which Search icon svg should we use?

georgewrmarshall replied:
Hey @amerkadicE, I think we just go with the current one for now `ui/components/ui/search-icon/search-icon.component.js`. Could you extend this one for your use case and use it for now? I've create a ticket to move this icon to it's appropriate locations https://github.com/MetaMask/metamask-extension/issues/13563

amerkadicE replied:
Done

georgewrmarshall commented on ui/pages/settings/index.scss:
```diff
@@ -11,22 +11,138 @@
+
+        > div {
+          &:hover {
+            background: --ui-1;
```
Update to correct syntax for css variables 
```suggestion
            background: var(--ui-1);
```

amerkadicE replied:
Done

georgewrmarshall commented on ui/pages/settings/index.scss:
```diff
@@ -11,22 +11,138 @@
+      }
+
+      &__list {
+        background: #fff;
```
Use css variables we should not be using any static hex / color values in the code base

```suggestion
        background: var(--white);
```

amerkadicE replied:
Done

georgewrmarshall commented on ui/pages/settings/index.scss:
```diff
@@ -11,22 +11,138 @@
+      &__close-button::after {
+        content: '\00D7';
+        font-size: 40px;
+        color: --ui-4;
```
use correct css variable syntax
```suggestion
        color: var(--ui-4);
```

amerkadicE replied:
Done

georgewrmarshall commented on ui/pages/settings/index.scss:
```diff
@@ -11,22 +11,138 @@
+          &__no-matching {
+            @include H6;
+
+            color: --ui-4;
```
Use correct css var syntax
```suggestion
            color: var(--ui-4);
```

amerkadicE replied:
Done

georgewrmarshall commented on ui/pages/settings/index.scss:
```diff
@@ -11,22 +11,138 @@
+            background-size: contain;
+            background-repeat: no-repeat;
+            background-position: center;
+            margin: 0 5px;
```
Adhere to 8px grid
```suggestion
            margin: 0 4px;
```

amerkadicE replied:
Done

georgewrmarshall reviewed (changes requested):
Looking good! Did a review of the css and also running into this interesting issues when searching "Security &"

<img width="472" alt="Screen Shot 2022-02-09 at 11 45 38 AM" src="https://user-images.githubusercontent.com/8112138/153107410-15a27531-af77-4b6f-b31c-bcddd3b8942f.png">

georgewrmarshall commented:
Also was playing around with css grid for the list items. I removed the margins on all items and updated it to this

```
.settings-page__header__search__list__item {
    transition: 200ms ease-in-out;
    display: grid;
    align-items: center;
    padding: 16px;
    border-top: 1px solid var(--ui-2);
    cursor: pointer;
    grid-template-columns: 16px minmax(20px, max-content) 8px auto;
    gap: 8px;
}
```

Which may make it a little more intuitive and like the intended figma designs

https://user-images.githubusercontent.com/8112138/153107648-3496971a-034d-445f-9f59-3e99066100d2.mov

You would have to fix the issue of the last item though

<img width="511" alt="Screen Shot 2022-02-09 at 12 26 15 PM" src="https://user-images.githubusercontent.com/8112138/153107700-25614bf2-6ec1-4e5d-8976-4ba2507085d6.png">

amerkadicE commented on ui/pages/settings/index.scss:
```diff
@@ -11,22 +11,138 @@
-    &__title {
-      @include H3;
+    @media screen and (max-width: $break-small) {
+      background: #f2f3f4;
```
Done

amerkadicE commented on ui/pages/settings/index.scss:
```diff
@@ -11,22 +11,138 @@
-    padding: 12px 24px;
-    align-items: center;
-    flex: 0 0 auto;
+    padding: 12px 24px 5px 24px;
```
Done

amerkadicE commented on ui/pages/settings/index.scss:
```diff
@@ -11,22 +11,138 @@
+      @media screen and (min-width: $break-large) {
+        position: absolute;
+        right: 57px;
+        top: 13px;
```
Done

danjm reviewed (dismissed):
LGTM

georgewrmarshall reviewed (changes requested):
Code looks good! But I ran into an issues that looks like it has been created because of some code being merged in recently 

<img width="369" alt="Screen Shot 2022-02-15 at 11 27 08 AM" src="https://user-images.githubusercontent.com/8112138/153973242-aef78d00-39f6-41f4-9ba1-a269f39557f2.png">


@dragana8 can you explain what the reason is behind the [z-index here](https://github.com/MetaMask/metamask-extension/blob/develop/ui/components/app/tab-bar/index.scss#L43-L44) it was merge in [this commit](https://github.com/MetaMask/metamask-extension/commit/47e2e377122d3aafa1602ca3749341a911115bbd)

dragana8 commented:
@georgewrmarshall I don't really remember but I told @amerkadicE to remove it if it's causing issues here and tab-bar is ok without it.

brad-decker commented:
I have tried to think of an architectural approach that doesn't require the registering of refs, and the sort of generative json structure that is proposed in this PR and cannot. This is largely because anything that I would suggest would require a significant overhaul of the settings page itself. I have tracked an issue for an idea that could potentially get us to a good place with search capabilities while also positioning future redesigns as easier on the settings screen: #13672

mohamadrezakheradmand commented:
> Code looks good! But I ran into an issues that looks like it has been created because of some code being merged in recently 
> 
> <img width="369" alt="Screen Shot 2022-02-15 at 11 27 08 AM" src="https://user-images.githubusercontent.com/8112138/153973242-aef78d00-39f6-41f4-9ba1-a269f39557f2.png">
> 
> 
> @dragana8 can you explain what the reason is behind the [z-index here](https://github.com/MetaMask/metamask-extension/blob/develop/ui/components/app/tab-bar/index.scss#L43-L44) it was merge in [this commit](https://github.com/MetaMask/metamask-extension/commit/47e2e377122d3aafa1602ca3749341a911115bbd)

georgewrmarshall reviewed (dismissed):
UI LGTM! Nice work!

georgewrmarshall reviewed (approved):
UI lgtm!
